internal/uvm/vsmb.go

```diff
@@ -203,13 +206,91 @@ func (uvm *UtilityVM) GetVSMBUvmPath(ctx context.Context, hostPath string, readO
 	if err != nil {
 		return "", err
 	}
-	return filepath.Join(share.guestPath, f), nil
+	return filepath.Join(share.GuestPath, f), nil
 }
 
-// getVSMBShareKey returns a string key which encapsulates the information that
-// is used to look up an existing VSMB share. If a share is being added, but
-// there is an existing share with the same key, the existing share will be used
-// instead (and its ref count incremented).
+var _ = (Cloneable)(&VSMBShare{})
+
+// GobEncode serializes the VSMBShare struct
+func (vsmb *VSMBShare) GobEncode() ([]byte, error) {
+	var buf bytes.Buffer
+	encoder := gob.NewEncoder(&buf)
+	errMsgFmt := "failed to encode VSMBShare: %s"
+	// encode only the fields that can be safely deserialized.
+	if err := encoder.Encode(vsmb.HostPath); err != nil {
+		return nil, fmt.Errorf(errMsgFmt, err)
+	}
+	if err := encoder.Encode(vsmb.Name); err != nil {
+		return nil, fmt.Errorf(errMsgFmt, err)
+	}
+	if err := encoder.Encode(vsmb.AllowedFiles); err != nil {
+		return nil, fmt.Errorf(errMsgFmt, err)
+	}
+	if err := encoder.Encode(vsmb.GuestPath); err != nil {
+		return nil, fmt.Errorf(errMsgFmt, err)
+	}
+	if err := encoder.Encode(vsmb.Options); err != nil {
+		return nil, fmt.Errorf(errMsgFmt, err)
+	}
+	return buf.Bytes(), nil
+}
+
+// GobDecode deserializes the VSMBShare struct into the struct on which this is called
+// (i.e the vsmb pointer)
+func (vsmb *VSMBShare) GobDecode(data []byte) error {
+	buf := bytes.NewBuffer(data)
+	decoder := gob.NewDecoder(buf)
+	errMsgFmt := "failed to decode VSMBShare: %s"
+	// fields should be decoded in the same order in which they were encoded.
+	if err := decoder.Decode(&vsmb.HostPath); err != nil {
+		return fmt.Errorf(errMsgFmt, err)
+	}
+	if err := decoder.Decode(&vsmb.Name); err != nil {
+		return fmt.Errorf(errMsgFmt, err)
+	}
+	if err := decoder.Decode(&vsmb.AllowedFiles); err != nil {
+		return fmt.Errorf(errMsgFmt, err)
+	}
+	if err := decoder.Decode(&vsmb.GuestPath); err != nil {
+		return fmt.Errorf(errMsgFmt, err)
+	}
+	if err := decoder.Decode(&vsmb.Options); err != nil {
+		return fmt.Errorf(errMsgFmt, err)
+	}
+	return nil
+}
+
+// Clone creates a clone of the VSMBShare `vsmb` and adds that clone to the uvm `vm`.  To
+// clone VSMB share we just need to add it into the config doc of that VM and increase the
+// vsmb counter.
+func (vsmb *VSMBShare) Clone(ctx context.Context, vm *UtilityVM, cd *cloneData) error {
+	cd.doc.VirtualMachine.Devices.VirtualSmb.Shares = append(cd.doc.VirtualMachine.Devices.VirtualSmb.Shares, hcsschema.VirtualSmbShare{
+		Name:         vsmb.Name,
+		Path:         vsmb.HostPath,
+		Options:      &vsmb.Options,
+		AllowedFiles: vsmb.AllowedFiles,
+	})
+	vm.vsmbCounter++
+
+	clonedVSMB := &VSMBShare{
+		vm:           vm,
+		HostPath:     vsmb.HostPath,
+		refCount:     1,
+		Name:         vsmb.Name,
+		Options:      vsmb.Options,
+		AllowedFiles: vsmb.AllowedFiles,
+		GuestPath:    vsmb.GuestPath,
+	}
+
+	vm.vsmbDirShares[vsmb.HostPath] = clonedVSMB
```

Comment:
This seems like it wouldn't work right if the share was in `vsmbFileShares` previously. Do we need to handle that case?

Reply:
I had actually fixed this in the 4th PR but for the sake of completeness I have added that fix here too.